### PR TITLE
3465 create new asset catalogue items via import spreadsheet

### LINKED ISSUE #3465: Create new asset catalogue items via import spreadsheet
## Is your feature request related to a problem? Please describe 👀

Asset managers want to be able to create new catalogue items in a `sub catalogue`.

## Describe the solution you'd like 🎁

The assets view should have an import button and a `Sub Catalogue` field added to the view.
image

### Acceptance Criteria

- [x] Adding catalogue items should only be possible from the central server
- [x] Able to filter/sort by `sub catalogue` name
- [x] Able to select which fields are shown in the asset list
- [x] Import should require the following fields
     - sub catalogue, code, manufacturer, model, class, category, and type
- [x] The import should validate that there are no duplicate codes
- [x] The import should validate that the same manufacturer and model doesn't already exist in the system

### Describe alternatives you've considered 💭
UI for editing asset items, which will take more time.

### Additional context 💌

To make the code simple for now if we just add a string field for `catalogue name`, and we'll only support creating new catalogue items via an import modal.

> This means that in future if we want to create asset catalogues as their own separate entities we'll need to migrate existing catalogue names into the new structure.

### Moneyworks Jobcode 🧰

OMSCC:AST

### PR DESCRIPTION
**Pull request by jmbrunskill**
Fixes #3465 

# 👩🏻‍💻 What does this PR do? 
 

- [x] Adds Import and Delete functionality for Asset Catalogue Items on Central Server
image

image

On non-central server the import & delete buttons just don't appear, permissions are only checked on the backend currently.

- [x] Adds `Sub Catalogues`
image

# 🧪 How has/should this change been tested? 

Start a central server.
e.g.`IS_CENTRAL_SERVER=true cargo run`

Go to Catalogue -> Assets

Try adding a new asset item via the `Import` Button
Try deleting the asset item

Check that...

- [ ] Only users with `Setup Assets` access can create and delete asset catalogue items
image
- [ ] If not on central server these actions can't be taken
- [ ] You can't delete an asset item that's used by an asset
- [ ] Imported Catalogue items are available on a remote site after sync
- [ ] Check the import example is correct and makes sense
- [ ] Check that errors are displayed correctly if an import fails

## 💌 Any notes for the reviewer?

I'm not sure if delete works via sync?

## 📃 Documentation

- [ ] Importing for asset items

### CODE REVIEW

File: client/packages/system/src/Asset/ImportCatalogueItem/UploadTab.tsx [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+  catalogueItemData?: AssetCatalogueItemFragment[];
+}
+
+// introduce new interface to accommodate dynamic keys of parsed result
+interface ParsedImport {
+  id: string;
+  [key: string]: string | undefined;
+}
+
+enum AssetColumn {
```
Comment (jmbrunskill) [inline client/packages/system/src/Asset/ImportCatalogueItem/UploadTab.tsx:38]: Don't love this approach but copied over from Asset Import

File: client/packages/system/src/Asset/utils.ts [hunk 2]
```diff
@@ -1,23 +1,34 @@
...
+}
+
+export const assetCatalogueItemsListToCsv = (
+  items: AssetCatalogueItemFragment[],
+  t: TypedTFunction<LocaleKey>
+) => {
+  const fields = assetCatalogueItemFields(t);
 
   const data = items.map(node => [
-    node.id,
```
Comment (jmbrunskill) [inline client/packages/system/src/Asset/utils.ts:17]: Deliberately excluding ID as we don't want it in imports, not sure if adds value in export either except for perhaps our own team?

File: server/graphql/asset_catalogue/Cargo.toml [hunk 3]
```diff
@@ -13,6 +13,7 @@ repository = { path = "../../repository" }
 service = { path = "../../service" }
 util = { path = "../../util" }
 graphql_core = { path = "../core" }
+graphql_types = { path = "../types" }
```
Comment (jmbrunskill) [inline server/graphql/asset_catalogue/Cargo.toml:16]: Sad to add a build dependancy, but didn't want to duplicate DeleteReponseType 😢  
That should maybe be moved to core but already more than enough changes in this PR.

File: client/packages/common/src/intl/locales/en/coldchain.json [hunk 4]
```diff
@@ -26,6 +26,7 @@
   "error.code-no-match": "{{field}} not found in catalogue",
   "error.unable-to-scan": "Unable to scan QR code",
   "error.duplicate-asset-number": "duplicate asset number in csv",
+  "error.invalid-field-value": "{{ value }} isn't a valid {{ field }}",
```
Comment (jmbrunskill) [inline client/packages/common/src/intl/locales/en/coldchain.json:29]: Is it ok for coldchain translations to be used in catalogue, or should I move some things around?

File: client/packages/system/src/Asset/utils.ts [hunk 5]
```diff
@@ -1,23 +1,34 @@
...
 import { Formatter } from '@common/utils';
+import { ImportRow, LineNumber } from './ImportCatalogueItem';
 
-export const assetCategoryListItemsToCsv = (
-  items: AssetCatalogueItemFragment[],
-  t: TypedTFunction<LocaleKey>
-) => {
-  const fields: string[] = [
-    'id',
+function assetCatalogueItemFields(t: TypedTFunction<LocaleKey>) {
```
Comment (mark-prins) [inline client/packages/system/src/Asset/utils.ts:6]: a function and not a const? 🤔

File: server/graphql/asset_catalogue/Cargo.toml (hunk 3 above)
Comment (mark-prins) [reply server/graphql/asset_catalogue/Cargo.toml:16]: Good point - moving to core would be better.. but not now!

**Review by mark-prins — CHANGES_REQUESTED** [review]
🎉  Woo! nice work @jmbrunskill - there's a lot in there! works nicely, and I've picked up an unrelated bug in testing, so this was helpful.

A few comments in the f/e but just one thing required to change

One showstopper - on postgres, this needs to be in the migrations:

```
ALTER TYPE activity_log_type ADD VALUE 'ASSET_CATALOGUE_ITEM_CREATED';
```

Would it be possible to have the error messages translated?

Screenshot 2024-04-10 at 8 50 53 AM

`client/packages/system/src/Asset/api/hooks/line/useAssets/index.ts` is an empty file. is it needed?

File: client/packages/system/src/Asset/ImportCatalogueItem/CatalogueItemImportModal.tsx [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+};
+
+export type LineNumber = {
+  lineNumber: number;
+};
+export const toInsertAssetItemInput = (
+  row: ImportRow
+): AssetCatalogueItemFragment => {
+  return {
+    __typename: 'AssetCatalogueItemNode',
```
[the remark this answers is not in the document]
Comment (jmbrunskill) [reply client/packages/system/src/Asset/ImportCatalogueItem/CatalogueItemImportModal.tsx:57]:
```
fragment AssetCatalogueItem on AssetCatalogueItemNode {
  assetCategoryId
  assetClassId
  assetTypeId
  code
  id
  manufacturer
  model
  subCatalogue
  assetClass {
    name
  }
  assetCategory {
    name
  }
  assetType {
    name
  }
}
```
Here's my fragment, so not sure why it has __typename required?

File: client/packages/system/src/Asset/ImportCatalogueItem/CatalogueItemImportModal.tsx [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+});
+
+export const AssetCatalogueItemImportModal: FC<AssetItemImportModalProps> = ({
+  isOpen,
+  onClose,
+}) => {
+  const t = useTranslation('coldchain');
+  const { success } = useNotification();
+  const { currentTab, onChangeTab } = useTabs(Tabs.Upload);
+  const [activeStep, setActiveStep] = useState(0);
```
[the remark this answers is not in the document]
Comment (jmbrunskill) [reply client/packages/system/src/Asset/ImportCatalogueItem/CatalogueItemImportModal.tsx:84]: Yeah I had similar thoughts, maybe I should at least create an issue for it?

File: client/packages/system/src/Asset/utils.ts (hunk 5 above)
Comment (jmbrunskill) [reply client/packages/system/src/Asset/utils.ts:6]: Can it be const with the translations?

**Comment by jmbrunskill** [comment]
> `client/packages/system/src/Asset/api/hooks/line/useAssets/index.ts` is an empty file. is it needed?

I don't seem to have this file, is it just on your computer?

**Comment by jmbrunskill** [comment]
> One showstopper - on postgres, this needs to be in the migrations:
> 
> ```
> ALTER TYPE activity_log_type ADD VALUE 'ASSET_CATALOGUE_ITEM_CREATED';
> ```

Oops, I did run postgres tests, but I see they had actually failed! Must have forgotten to wait for them to finish...


Will work on the Translation for Errors and error message on delete tomorrow

File: client/packages/system/src/Asset/utils.ts (hunk 5 above)
Comment (mark-prins) [reply client/packages/system/src/Asset/utils.ts:6]: yeah, should be ok

**Review by mark-prins — APPROVED** [review]

**Comment by jmbrunskill** [comment]
> > I don't seem to have this file, is it just on your computer?
> 
> I didn't have it either! it was showing in the GH diff tho 🤷 can just ignore in that case

I remember now, this was a emtpy file I deleted in this PR.

**Pull request merged**